From d7507140eab4bf9f1ae3d3721d7803fff9b0c175 Mon Sep 17 00:00:00 2001
From: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Fri, 23 Sep 2022 17:38:27 -0300
Subject: [PATCH 04/55] emit: Don't mark a enum with nr_members == 0 as
 printed, its just a fwd decl

And with this now we can do a 'pahole --compile' om vmlinux from DWARF
info and it works :-)

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 dwarves_emit.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/dwarves_emit.c b/dwarves_emit.c
index 304d587..7d90d42 100644
--- a/dwarves_emit.c
+++ b/dwarves_emit.c
@@ -121,7 +121,13 @@ static int enumeration__emit_definitions(struct tag *tag,
 
 	enumeration__fprintf(tag, conf, fp);
 	fputs(";\n", fp);
-	type_emissions__add_definition(emissions, etype);
+
+	// See comment on enumeration__fprintf(), it seems this happens with DWARF as well
+	// or BTF doesn't have type->declaration set because DWARF didn't have it set.
+	// But we consider type->nr_members == 0 as just a forward declaration, so don't
+	// mark it as defined because we may need it to __really__ printf it later.
+	if (etype->nr_members != 0)
+		type_emissions__add_definition(emissions, etype);
 	return 1;
 }
 
-- 
2.40.0.rc2

